"ValueError: Cannot parse '470' to resistance" (populate)

Hello,
I'm trying to use the res_band feature with populate.
I get this error:

Traceback (most recent call last):
  File "/home/david/.local/bin/pcbdraw", line 8, in <module>
    sys.exit(run())
  File "/home/david/.local/lib/python3.8/site-packages/click/core.py", line 1128, in __call__
    return self.main(*args, **kwargs)
  File "/home/david/.local/lib/python3.8/site-packages/click/core.py", line 1053, in main
    rv = self.invoke(ctx)
  File "/home/david/.local/lib/python3.8/site-packages/click/core.py", line 1659, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/david/.local/lib/python3.8/site-packages/click/core.py", line 1395, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/david/.local/lib/python3.8/site-packages/click/core.py", line 754, in invoke
    return __callback(*args, **kwargs)
  File "/home/david/.local/lib/python3.8/site-packages/pcbdraw/populate.py", line 300, in populate
    parsed_content = generate_images(parsed_content, board, prepare_params(header["params"]),
  File "/home/david/.local/lib/python3.8/site-packages/pcbdraw/populate.py", line 193, in generate_images
    generate_image(boardfilename, x["side"], x["components"],
  File "/home/david/.local/lib/python3.8/site-packages/pcbdraw/populate.py", line 212, in generate_image
    plot.main(args=plot_args)
  File "/home/david/.local/lib/python3.8/site-packages/click/core.py", line 1053, in main
    rv = self.invoke(ctx)
  File "/home/david/.local/lib/python3.8/site-packages/click/core.py", line 1395, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/david/.local/lib/python3.8/site-packages/click/core.py", line 754, in invoke
    return __callback(*args, **kwargs)
  File "/home/david/.local/lib/python3.8/site-packages/pcbdraw/ui.py", line 174, in plot
    image = plotter.plot()
  File "/home/david/.local/lib/python3.8/site-packages/pcbdraw/plot.py", line 1053, in plot
    plotter.render(self)
  File "/home/david/.local/lib/python3.8/site-packages/pcbdraw/plot.py", line 812, in render
    plotter.walk_components(invert_side=False, callback=self._append_component)
  File "/home/david/.local/lib/python3.8/site-packages/pcbdraw/plot.py", line 1084, in walk_components
    callback(lib, name, ref, value, pos)
  File "/home/david/.local/lib/python3.8/site-packages/pcbdraw/plot.py", line 840, in _append_component
    ret = self._create_component(lib, name, ref, value)
  File "/home/david/.local/lib/python3.8/site-packages/pcbdraw/plot.py", line 890, in _create_component
    self._apply_resistor_code(component_element, id_prefix, ref, value)
  File "/home/david/.local/lib/python3.8/site-packages/pcbdraw/plot.py", line 912, in _apply_resistor_code
    res, tolerance = self._get_resistance_from_value(value)
  File "/home/david/.local/lib/python3.8/site-packages/pcbdraw/plot.py", line 947, in _get_resistance_from_value
    res = read_resistance(value_l[0])
  File "/home/david/.local/lib/python3.8/site-packages/pcbdraw/unit.py", line 39, in read_resistance
    raise ValueError(f"Cannot parse '{value}' to resistance")
ValueError: Cannot parse '470' to resistance

If I understand correctly, the value '470' is retrieved from the .kicad_pcb file.
The script parses correctly an other value (100k) and generates correctly the corresponding picture.

Also, in this case I use the default resistor footprint. Not my custom one (see other issue #116 )

Thank you.

Thank you very much @Electro707 !
However, I get this error when I try either plot or populate commands:

david@b101p02:~/tmp/test/PcbDraw$ python -m pcbdraw.ui populate ~/elec/simple\ vca/1.0/populate.txt ~/elec/simple\ vca/1.0/img/
wx._core.wxAssertionError: C++ assertion "false" failed at /build/kicad-YdG4kW/kicad-6.0.8-0/kicad/common/layer_id.cpp(180) in LayerName(): Unknown layer ID -1

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/home/david/tmp/test/PcbDraw/pcbdraw/ui.py", line 312, in <module>
    run()
  File "/home/david/.local/lib/python3.8/site-packages/click/core.py", line 1128, in __call__
    return self.main(*args, **kwargs)
  File "/home/david/.local/lib/python3.8/site-packages/click/core.py", line 1053, in main
    rv = self.invoke(ctx)
  File "/home/david/.local/lib/python3.8/site-packages/click/core.py", line 1659, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/david/.local/lib/python3.8/site-packages/click/core.py", line 1395, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/david/.local/lib/python3.8/site-packages/click/core.py", line 754, in invoke
    return __callback(*args, **kwargs)
  File "/home/david/tmp/test/PcbDraw/pcbdraw/populate.py", line 300, in populate
    parsed_content = generate_images(parsed_content, board, prepare_params(header["params"]),
  File "/home/david/tmp/test/PcbDraw/pcbdraw/populate.py", line 193, in generate_images
    generate_image(boardfilename, x["side"], x["components"],
  File "/home/david/tmp/test/PcbDraw/pcbdraw/populate.py", line 212, in generate_image
    plot.main(args=plot_args)
  File "/home/david/.local/lib/python3.8/site-packages/click/core.py", line 1053, in main
    rv = self.invoke(ctx)
  File "/home/david/.local/lib/python3.8/site-packages/click/core.py", line 1395, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/david/.local/lib/python3.8/site-packages/click/core.py", line 754, in invoke
    return __callback(*args, **kwargs)
  File "/home/david/tmp/test/PcbDraw/pcbdraw/ui.py", line 174, in plot
    image = plotter.plot()
  File "/home/david/tmp/test/PcbDraw/pcbdraw/plot.py", line 1053, in plot
    plotter.render(self)
  File "/home/david/tmp/test/PcbDraw/pcbdraw/plot.py", line 674, in render
    self._plotter.execute_plot_plan(to_plot)
  File "/home/david/tmp/test/PcbDraw/pcbdraw/plot.py", line 1207, in execute_plot_plan
    pctl.OpenPlotfile(action.name, pcbnew.PLOT_FORMAT_SVG, action.name)
  File "/usr/lib/python3/dist-packages/pcbnew.py", line 7586, in OpenPlotfile
    return _pcbnew.PLOT_CONTROLLER_OpenPlotfile(self, aSuffix, aFormat, aSheetDesc)
SystemError: <built-in function PLOT_CONTROLLER_OpenPlotfile> returned a result with an error set


I'm not familiar with python, so I have no idea if I'm using it the right way...
(`python -m pcbdraw.ui` doesn't generate this error)

--- a/pcbdraw/plot.py
+++ b/pcbdraw/plot.py
@@ -951,20 +951,20 @@ class PlotComponents(PlotInterface):
 
     def _get_resistance_from_value(self, value: str) -> Tuple[Decimal, str]:
         res, tolerance = None, "5%"
+        value_l = value.split(" ", maxsplit=1)
         try:
-            value_l = value.split(" ", maxsplit=1)
             res = read_resistance(value_l[0])
-            if len(value_l) > 1:
-                t_string = value_l[1].strip().replace(" ", "")
-                if "%" in t_string:
-                    s = self._plotter.get_style("tht-resistor-band-colors")
-                    if not isinstance(s, dict):
-                        raise RuntimeError(f"Invalid style specified, tht-resistor-band-colors should be dictionary, got {type(s)}")
-                    if t_string.strip() not in s:
-                        raise UserWarning(f"Invalid resistor tolerance {value_l[1]}")
-                    tolerance = t_string
-        except decimal.InvalidOperation:
-            raise UserWarning(f"Invalid value {value}") from None
+        except ValueError:
+            raise UserWarning(f"Invalid resistor value {value_l[0]}")
+        if len(value_l) > 1:
+            t_string = value_l[1].strip().replace(" ", "")
+            if "%" in t_string:
+                s = self._plotter.get_style("tht-resistor-band-colors")
+                if not isinstance(s, dict):
+                    raise RuntimeError(f"Invalid style specified, tht-resistor-band-colors should be dictionary, got {type(s)}")
+                if t_string.strip() not in s:
+                    raise UserWarning(f"Invalid resistor tolerance {value_l[1]}")
+                tolerance = t_string
         return res, tolerance
 
 

--- a/pcbdraw/ui.py
+++ b/pcbdraw/ui.py
@@ -197,7 +197,7 @@ def build_plot_components(remap: str, highlight: List[str], filter: Optional[Lis
         key, value = tuple(mapping.split(":"))
         resistor_values[key] = ResistorValue(value=value)
     for ref in resistor_flip:
-        field = resistor_values.get(key, ResistorValue())
+        field = resistor_values.get(ref, ResistorValue())
         field.flip_bands = True
         resistor_values[ref] = field
 

--- a/pcbdraw/unit.py
+++ b/pcbdraw/unit.py
@@ -15,25 +15,35 @@ def erase(string: str, what: List[str]) -> str:
 def read_resistance(value: str) -> Decimal:
     """
     Given a string, try to parse resistance and return it as Ohms (Decimal)
+
+    This function can raise a ValueError if the value is invalid
     """
     p_value = erase(value, ["Ω", "Ohms", "Ohm"]).strip()
     p_value = p_value.replace(" ", "") # Sometimes there are spaces after decimal place
     unit_prefixes = {
-        "m": [Decimal(1e-3), Decimal(1e-6)],
-        "K": [Decimal(1e3), Decimal(1)],
-        "k": [Decimal(1e3), Decimal(1)],
-        "M": [Decimal(1e6), Decimal(1e3)],
-        "G": [Decimal(1e9), Decimal(1e6)]
+        "m": Decimal('1e-3'),
+        "R": Decimal('1'),
+        "K": Decimal('1e3'),
+        "k": Decimal('1e3'),
+        "M": Decimal('1e6'),
+        "G": Decimal('1e9')
     }
     try:
         numerical_value = None
         for prefix, table in unit_prefixes.items():
             if prefix in p_value:
-                split = [Decimal(x) if x != "" else Decimal(0) for x in p_value.split(prefix)]
-                numerical_value = split[0] * table[0] + split[1] * table[1]
+                # Example: 4k7 will have the 4 converted to Decimal(4) and 7 to Decimal(0.7)
+                # Then each gets multiplied by the factor and added, so 4000 + 700
+                # This method ensures that 4k7 and 4k700 for example yields the same result
+                split = p_value.split(prefix)
+                n_whole = Decimal(split[0]) if split[0] != "" else Decimal(0)
+                n_dec = Decimal('.'+split[1]) if split[1] != "" else Decimal(0)
+                numerical_value = n_whole * table + n_dec * table
                 break
-        if numerical_value is not None:
-            return numerical_value
+        if numerical_value is None:
+            # If this fails, a decimal.InvalidOperation is raised which is handled by the Exception catch
+            numerical_value = Decimal(p_value)
+        return numerical_value
     except Exception:
         pass
     raise ValueError(f"Cannot parse '{value}' to resistance")

